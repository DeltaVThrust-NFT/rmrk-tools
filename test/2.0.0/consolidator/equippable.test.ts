import { Consolidator } from "../../../src/rmrk2.0.0";
import {
  createNftClassMock,
  getBlockCallsMock,
  getBobKey,
  getRemarksFromBlocksMock,
  createBaseMock,
} from "../mocks";
import { cryptoWaitReady } from "@polkadot/util-crypto";

beforeAll(async () => {
  return await cryptoWaitReady();
});

describe("rmrk2.0.0 Consolidator: EQUIPPABLE", () => {
  const getSetupRemarks = () => [
    ...getBlockCallsMock(createNftClassMock().create()),
    ...getBlockCallsMock(createBaseMock().base()),
  ];

  it("Add new Class id to Base slot equippable", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(
        createBaseMock(3).equippable(
          "background",
          [createNftClassMock().id],
          "+"
        )
      ),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("Remove new Class id to Base slot equippable", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(
        createBaseMock(3).equippable("backpack", [createNftClassMock().id], "-")
      ),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("Replace a Class id to Base slot equippable", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(createBaseMock(3).equippable("backpack", ["*"], "")),
    ]);
    const consolidator = new Consolidator();
    expect(await consolidator.consolidate(remarks)).toMatchSnapshot();
  });

  it("Should invalidate EQUIPPABLE if base parts slot is missing", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(
        createBaseMock(3).equippable("test", [createNftClassMock().id], "+")
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult.invalid[0].message).toEqual(
      "[EQUIPPABLE] Attempting to change equippable on non-existant part with a slot id test"
    );
  });

  it("Should invalidate EQUIPPABLE if base is missing", async () => {
    const remarks = getRemarksFromBlocksMock([
      ...getSetupRemarks(),
      ...getBlockCallsMock(
        createBaseMock(5).equippable("test", [createNftClassMock().id], "+")
      ),
    ]);
    const consolidator = new Consolidator();
    const consolidatedResult = await consolidator.consolidate(remarks);
    expect(consolidatedResult.invalid[0].message).toEqual(
      "[EQUIPPABLE] Attempting to change equippable on non-existant NFT base-5-KBASE777"
    );
  });
});
