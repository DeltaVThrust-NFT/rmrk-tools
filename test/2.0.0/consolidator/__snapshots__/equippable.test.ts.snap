// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rmrk2.0.0 Consolidator: EQUIPPABLE Add new Collection id to Base slot equippable 1`] = `
Object {
  "bases": Array [
    Object {
      "block": 3,
      "changes": Array [],
      "id": "base-3-KBASE777",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "parts": Array [
        Object {
          "equippable": Array [
            "d43593c715a56da27d-KANARIABIRDS",
          ],
          "id": "background",
          "type": "slot",
          "z": 0,
        },
        Object {
          "equippable": Array [
            "d43593c715a56da27d-KANARIABIRDS",
          ],
          "id": "backpack",
          "type": "slot",
          "z": 1,
        },
        Object {
          "id": "tail",
          "src": "ipfs://ipfs/QmcEuigDVCScMLs2dcrJ8qU4Q265xGisUyKeYdnnGFn6AE/var3_tail.svg",
          "type": "fixed",
          "z": 2,
        },
      ],
      "symbol": "KBASE777",
      "themes": Object {
        "themeOne": Object {
          "_inherit": true,
          "primary": "#fff",
        },
      },
      "type": "svg",
    },
  ],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  ],
  "invalid": Array [],
  "nfts": Array [],
}
`;

exports[`rmrk2.0.0 Consolidator: EQUIPPABLE Remove new Collection id to Base slot equippable 1`] = `
Object {
  "bases": Array [
    Object {
      "block": 3,
      "changes": Array [],
      "id": "base-3-KBASE777",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "parts": Array [
        Object {
          "equippable": Array [],
          "id": "background",
          "type": "slot",
          "z": 0,
        },
        Object {
          "equippable": Array [],
          "id": "backpack",
          "type": "slot",
          "z": 1,
        },
        Object {
          "id": "tail",
          "src": "ipfs://ipfs/QmcEuigDVCScMLs2dcrJ8qU4Q265xGisUyKeYdnnGFn6AE/var3_tail.svg",
          "type": "fixed",
          "z": 2,
        },
      ],
      "symbol": "KBASE777",
      "themes": Object {
        "themeOne": Object {
          "_inherit": true,
          "primary": "#fff",
        },
      },
      "type": "svg",
    },
  ],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  ],
  "invalid": Array [],
  "nfts": Array [],
}
`;

exports[`rmrk2.0.0 Consolidator: EQUIPPABLE Replace a Collection id to Base slot equippable 1`] = `
Object {
  "bases": Array [
    Object {
      "block": 3,
      "changes": Array [],
      "id": "base-3-KBASE777",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "parts": Array [
        Object {
          "equippable": Array [],
          "id": "background",
          "type": "slot",
          "z": 0,
        },
        Object {
          "equippable": "*",
          "id": "backpack",
          "type": "slot",
          "z": 1,
        },
        Object {
          "id": "tail",
          "src": "ipfs://ipfs/QmcEuigDVCScMLs2dcrJ8qU4Q265xGisUyKeYdnnGFn6AE/var3_tail.svg",
          "type": "fixed",
          "z": 2,
        },
      ],
      "symbol": "KBASE777",
      "themes": Object {
        "themeOne": Object {
          "_inherit": true,
          "primary": "#fff",
        },
      },
      "type": "svg",
    },
  ],
  "collections": Array [
    Collection {
      "block": 2,
      "changes": Array [],
      "id": "d43593c715a56da27d-KANARIABIRDS",
      "issuer": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
      "max": 0,
      "metadata": "https://some.url",
      "symbol": "KANARIABIRDS",
    },
  ],
  "invalid": Array [],
  "nfts": Array [],
}
`;
